src/babel-serializer/geometry/obj.cc

```diff
@@ -1,2 +1,417 @@
+                {
+                    auto const vertex_idx = parse_int(idx);
+                    if (vertex_idx < 0)
+                        e.vertex_idx = int(geometry.vertices.size()) + vertex_idx;
+                    else
+                        e.vertex_idx = vertex_idx - 1;
+                }
+                else if (i == 1)
+                {
+                    auto const texture_idx = parse_int(idx);
+                    if (texture_idx < 0)
+                        e.tex_coord_idx = int(geometry.vertices.size()) + texture_idx;
+                    else
+                        e.tex_coord_idx = texture_idx - 1;
+                }
+                else
+                {
+                    on_error(data, cc::as_byte_span(s), "unable to parse line: unknown line segment format", severity::error);
+                }
+                ++i;
+            }
+            geometry.line_entries.push_back(e);
+        }
+
+        auto const entries_count = int(geometry.line_entries.size()) - entries_start;
+        if (entries_count < 2)
+        {
+            on_error(data, cc::as_byte_span(s), "unable to parse line: a line segment must contain at least two points", severity::error);
+        }
+        geometry.lines.push_back({entries_start, entries_count});
+    };
+
+    auto const handle_previous_groups = [&]() {
+        auto const points_count = int(geometry.points.size()) - group_point_start;
+        auto const lines_count = int(geometry.lines.size()) - group_line_start;
+        auto const faces_count = int(geometry.faces.size()) - group_face_start;
+
+        typename babel::obj::geometry<ScalarT>::group g;
+        if (points_count > 0)
+        {
+            g.points_start = group_point_start;
+            g.points_count = points_count;
+        }
+        if (lines_count > 0)
+        {
+            g.lines_start = group_line_start;
+            g.lines_count = lines_count;
+        }
+        if (faces_count > 0)
+        {
+            g.faces_start = group_face_start;
+            g.faces_count = faces_count;
+        }
+        if (points_count > 0 || lines_count > 0 || faces_count > 0)
+            for (auto const group : active_groups)
+            {
+                g.name = group;
+                geometry.groups.push_back(g);
+            }
+    };
+
+    auto const parse_groups = [&](cc::string_view s) {
+        CC_ASSERT(s[0] == 'g');
+        CC_ASSERT(cc::is_space(s[1]));
+
+        handle_previous_groups();
+        group_point_start = int(geometry.points.size());
+        group_line_start = int(geometry.lines.size());
+        group_face_start = int(geometry.faces.size());
+
+        // now parse
+        active_groups.clear();
+        for (auto const g : s.subview(2).split(cc::is_space, cc::split_options::skip_empty))
+            active_groups.push_back(g);
+    };
+
+    while (has_more_lines())
+    {
+        auto const line = next_line();
+        if (line.empty())
+            continue;
```

Comment:
An idea:
```
for (auto line : data_as_string_view.split(`\n`))
{
   line = line.trim();
   if (line.empty() || line.starts_with('#'))
      continue;

   ... rest
}
```

Reply:
Yeah this is cleaner, but I think it requires a `trim_comments` or something like that, because I believe it is valid to have a comment start mid-line.

Reply:
`line = *line.split('#').begin();` before the trim should do the trick (also works if no comments are there)